### DISCUSSION TITLE
Initiation of a displacement field

### DISCUSSION THREAD

**Discussion by p-rapan** [thelfer/tfel#359] [Q&A]
Hello everyone, 

I am using CA along with MFront in a HM coupling. I remember from previous developments that mfront was not getting the initial stress fields  sent to it by CA and I had to manually insert this into MFront. 
So my questions:
1. Now I except of the initial stress fields, I also have a displacement field (NOEU_DEPL_R) . Should I go proceed into any modifications?
2. How is it possible to insert into MFront my effective stresses?

Thanks a lot.

**Comment by thelfer** [comment]
Hi @Padrp

1. Displacement is handled by `code_aster`, `MFront` can't handle this
2. Yes, it depends how you want to interact with `code_aster`. If the fluid pressure is computed on the `code_aster` side, you must pass it to the behaviour via an external state variable. You probably need to return the total stress from the behaviour. @SRaude, would you comment on this ?

**Reply by SRaude** [reply]
Hi @Padrp and @thelfer,

1. First, if you define an initial stress field via `Code_Aster`, this field is automatically transferred at each Gauss point in the `MFront` constitutive library, so you won't have any problem with stresses ; but the problem is that the elastic strain tensor is part of the state variable's vector, so you will need to equilibrate those two fields. If you deal with linear elasticity, you can do that by defining an `@AuxiliaryStateVariable Inst` and initialising the elastic strain in the `@InitLocalVariables` block

    ```
    @InitLocalVariables{
	    if (Inst==time{0}) {
            const Stensor4 Se=invert(De);
            eel=(Se*sig);
        }
     }
    ```
    Then, you can update `Inst` in the corresponding block by `Inst+=dt`.

2. Second, MFront _only deals with_ stresses and strains (which are conjugated quantities). If you want to define an initial displacement field in Code_Aster, you can do that in the same way as for your initial stress field. There is no direct interaction between displacements defined in `Code_Aster` and `MFront` libraries.
3. Finally, when you deal with THM behaviours in `Code_Aster`, the stress transferred to `MFront` is always the effective stress ; so you won't have to do anything, `sig` is already the effective stress.

I hope this helps you,

Simon

**Reply by p-rapan** [reply]
Hello and thanks a lot for your replies. 

My mechanical law (MFront) comprises of a simple Drucker Prager elasto-perfectly plastic constitutive law that was created in order to be able to imitate the DP that already exists in code aster. For just D_PLAN models, I manage to get results close enough for the two behaviours (My user defined DP  and the one from the Code Aster). 

After that I jumped to the HM modelling, where normally I was expecting them to converge similarly just as in the pure mechanical case. In this case I don't manage to see the MFront converging properly (fluctuating convergence). So my mind goes naturally to the definition of the initial fields and how MFront understands them. 

In the HM modelling I will have something like: 

NOM_CMP=('SIXX','SIYY','SIZZ','SIXY','SIPXX','SIPYY','SIPZZ','SIPXY'), VALE=(-2.5e6,-4.5e6,-0.5e6,0.0,-pw,-pw, -pw, 0.0) inside the CREA_CHAMP. 

I am not very sure of how MFront understands this field and if I have to change sth inside the script. 
Your help is valuable at this point in my research. 
Thanks in advance.

**Reply by SRaude** [reply]
Hi @Padrp,

I think (but i'm not sure because I've never tried to use MFront with Code_Aster in HM conditions) that your initial stress field wil be correctly read by MFront. You can write a "cout" in your source file to ensure that this is the case. Differences in numerical convergence can come from many things, especially when you deal with multiphysic problems.

Simon.

**Reply by thelfer** [reply]
Hi @Padrp,

Beware that MFront uses the elastic strain in most cases and generally discards the stress given by `code_aster`. So if you don't initialize the elastic strain properly, you may have troubles.

Thomas

**Reply by p-rapan** [reply]
I hope you are all fine. 

I am attaching my mfront file. I coded a Drucker-Prager law to imitate the one that exists already in Code Aster.
Mechanically it works fine, when I use it in HM however not. 
DP.txt

It seems to be reading the pore pressure properly, but there must be something that I don't see. 
I would like your view on it and if I am missing something. Have been working with MFront and Code Aster quite a few time, managing the make it work will push my work a lot. 

Thank you in advance for all the help.

**Reply by thelfer** [reply]
@Padrp. Sorry for the delay in the anwser. I was hoping that some else could answer. To be honest, I have no experience with HM in `code_aster`. Have you tried the code_aster forums ?

**Reply by SRaude** [reply]
@Padrp I'm sorry but I can't see your file on my phone. I'm going to try to see it tomorrow on my PC, and answer you ASAP.

**Reply by SRaude** [reply]
@Padrp, I'm sorry but I still can't see your file. Even on my computer. Maybe it could be more efficient if you directly send me your files (tar archive or link to download it) by email at simon.raude@protonmail.com, so I'll be able to have a look on your problems with Code_Aster (by trying to run your computations on my configuration).

When you run some computations with MFront + Code_Aster, you have to pay attention to the numerical residual errors you use for the resolution of the local constitutive model and the one for the global equilibrium. The default value for RESI_INTE_MAXI (or RESI_INTE_RELA, depending on how you call your MFront constitutive model inside Code_Aster) is typically not small enough for MFront constitutive models ; so you have to take something between 1e-14 and 1e-10 (and the default value is 1e-7 I think). Furthermore, in HM coupled problems, the Newton-Raphson algorithm  for the global equilibrium has to deal with pressures and displacements, so you have to take RESI_GLOB_RELA small enough to ensure a real convergence ; typical values are 1e-10 to 1e-8 ; and the default value is 1e-6. If your local constitutive model is robust, typical values are RESI_INTE_MAXI=1e-12 and RESI_GLOB_RELA=1e-8.
